### PR TITLE
A default implementation for Filterable

### PR DESCRIPTION
**Pull request by spacekitteh**

### CODE REVIEW

**Comment by fumieval** [comment]
I'd like to avoid this kind of approach, because users might accidentally pick up this instance if they forget to import a more efficient instance they intended to use. Overlapping instances are controversial because whether to import a module can change the behaviour without any errors or warnings.

**Comment by spacekitteh** [comment]
But then they're dealing with orphan instances, unless the module defining the type implements it anyway.

**Comment by treeowl** [comment]
I also oppose this approach. Better, I think, to offer a function to use to define the method and a `newtype` to use with `DerivingVia`.

**Comment by spacekitteh** [comment]
I'll do that, then.

File: src/Data/Witherable.hs [hunk 1]
```diff
@@ -273,6 +274,20 @@ hashNub :: (Witherable t, Eq a, Hashable a) => t a -> t a
...
 {-# INLINE hashNub #-}
 
+-- | A default implementation for 'mapMaybe'.
+mapMaybeDefault :: (Foldable f, Alternative f) => (a -> Maybe b) -> f a -> f b
+mapMaybeDefault p = foldr (\x xs -> case p x of
+    Just a -> pure a <|> xs
+    _ -> xs) empty
+{-# INLINABLE mapMaybeDefault #-}
+
+newtype FilterableWrapper f a = FilterableWrapper {unWrapFilterable :: f a}
```
Comment (fumieval) [inline src/Data/Witherable.hs]: Minor: I think the convention is to `unwrap`, with the name reflecting the underlying structure (WrappedAlternative or something).  http://hackage.haskell.org/package/base-4.12.0.0/docs/src/Control.Applicative.html#WrappedMonad

Comment (spacekitteh) [reply src/Data/Witherable.hs]: Fixed :)

**Comment by fumieval** [comment]
Thanks.

**Pull request merged**